File: src/player/PlayerController.js
import * as THREE from 'three';
import * as CANNON from 'cannon-es';

export class PlayerController {
    constructor(scene, world, inputManager, stats) {
        this.scene = scene;
        this.world = world;
        this.input = inputManager;
        this.stats = stats;

        this.health = stats.maxHealth;
        this.armor = 0;
        this.stamina = stats.maxStamina;

        this.setupCamera();
        this.setupPhysics();
        
        this.velocity = new THREE.Vector3();
        this.canJump = false;
    }

    setupCamera() {
        this.camera = new THREE.PerspectiveCamera(
            75,
            window.innerWidth / window.innerHeight,
            0.1,
            1000
        );
        this.camera.position.set(0, 1.6, 0);
    }

    setupPhysics() {
        const shape = new CANNON.Sphere(0.5);
        this.body = new CANNON.Body({
            mass: 80,
            shape: shape,
            position: new CANNON.Vec3(0, 2, 0),
            linearDamping: 0.9
        });
        this.world.addBody(this.body);

        this.body.addEventListener('collide', (e) => {
            const contact = e.contact;
            if (contact.bi === this.body || contact.bj === this.body) {
                this.canJump = true;
            }
        });
    }

    update(deltaTime) {
        this.handleMovement(deltaTime);
        this.handleRotation(deltaTime);
        this.handleStamina(deltaTime);
        this.updateCameraPosition();
    }

    handleMovement(deltaTime) {
        const moveInput = this.input.getMovementInput();
        
        let speed = this.stats.walkSpeed;
        if (this.input.isSprinting() && this.stamina > 0) {
            speed = this.stats.sprintSpeed;
        } else if (this.input.isCrouching()) {
            speed = this.stats.crouchSpeed;
        }

        const forward = new THREE.Vector3(0, 0, -1);
        forward.applyQuaternion(this.camera.quaternion);
        forward.y = 0;
        forward.normalize();

        const right = new THREE.Vector3(1, 0, 0);
        right.applyQuaternion(this.camera.quaternion);
        right.y = 0;
        right.normalize();

        const moveDir = new THREE.Vector3();
        moveDir.addScaledVector(forward, -moveInput.z);
        moveDir.addScaledVector(right, moveInput.x);
        moveDir.normalize();

        this.body.velocity.x = moveDir.x * speed;
        this.body.velocity.z = moveDir.z * speed;

        if (this.input.isJumping() && this.canJump) {
            this.body.velocity.y = this.stats.jumpForce;
            this.canJump = false;
        }
    }

    handleRotation(deltaTime) {
        const lookInput = this.input.getLookInput();
        
        // Rotate body left/right
        this.body.quaternion.setFromEuler(new CANNON.Euler(0, this.camera.rotation.y - lookInput.x * this.stats.mouseSensitivity, 0));
        this.camera.rotation.y -= lookInput.x * this.stats.mouseSensitivity;
        
        // Rotate camera up/down
        this.camera.rotation.x -= lookInput.y * this.stats.mouseSensitivity;
        this.camera.rotation.x = Math.max(-Math.PI / 2, Math.min(Math.PI / 2, this.camera.rotation.x));
    }

    handleStamina(deltaTime) {
        if (this.input.isSprinting() && this.canJump) {
            this.stamina -= this.stats.sprintStaminaCost * deltaTime;
        } else {
            this.stamina += this.stats.staminaRegenRate * deltaTime;
        }
        this.stamina = Math.max(0, Math.min(this.stats.maxStamina, this.stamina));
    }

    updateCameraPosition() {
        this.camera.position.copy(this.body.position);
        this.camera.position.y += 0.6;
    }

    takeDamage(damage) {
        if (this.armor > 0) {
            const armorDamage = Math.min(damage * 0.5, this.armor);
            this.armor -= armorDamage;
            damage -= armorDamage;
        }
        this.health -= damage;
        this.health = Math.max(0, this.health);
        
        if (this.health <= 0) {
            this.die();
        }
    }

    die() {
        console.log('Player died!');
    }

    getCamera() {
        return this.camera;
    }
}

File: Assets/Scripts/Player/PlayerController.cs
using UnityEngine;
using NotebookWar.Data;
using NotebookWar.Input;
using NotebookWar.UI;

namespace NotebookWar.Player
{
    [RequireComponent(typeof(CharacterController))]
    public class PlayerController : MonoBehaviour
    {
        [SerializeField] private PlayerStats stats;
        [SerializeField] private Transform cameraTransform;
        [SerializeField] private float mouseSensitivity = 2f;

        private CharacterController controller;
        private IInputProvider input;
        private Vector3 velocity;
        private bool isGrounded;
        private bool isCrouching;
        private float verticalRotation;

        private void Awake()
        {
            controller = GetComponent<CharacterController>();
            input = new KeyboardMouseInput();
            Cursor.lockState = CursorLockMode.Locked;
        }

        private void Update()
        {
            HandleMovement();
            HandleRotation();
            HandleStamina();
        }

        private void HandleMovement()
        {
            isGrounded = controller.isGrounded;

            if (isGrounded && velocity.y < 0)
                velocity.y = -2f;

            Vector2 moveInput = input.GetMovementInput();
            Vector3 move = transform.right * moveInput.x + transform.forward * moveInput.y;

            float speed = stats.walkSpeed;
            if (input.GetSprintInput() && stats.currentStamina > 0)
                speed = stats.sprintSpeed;
            else if (input.GetCrouchInput())
                speed = stats.crouchSpeed;

            controller.Move(move * speed * Time.deltaTime);

            if (input.GetJumpInput() && isGrounded)
                velocity.y = Mathf.Sqrt(stats.jumpForce * -2f * Physics.gravity.y);

            velocity.y += Physics.gravity.y * Time.deltaTime;
            controller.Move(velocity * Time.deltaTime);
        }

        private void HandleRotation()
        {
            Vector2 lookInput = input.GetLookInput();

            transform.Rotate(Vector3.up * lookInput.x * mouseSensitivity);

            verticalRotation -= lookInput.y * mouseSensitivity;
            verticalRotation = Mathf.Clamp(verticalRotation, -90f, 90f);
            cameraTransform.localRotation = Quaternion.Euler(verticalRotation, 0f, 0f);
        }

        private void HandleStamina()
        {
            if (input.GetSprintInput() && isGrounded)
                stats.currentStamina -= stats.sprintStaminaCost * Time.deltaTime;
            else
                stats.currentStamina += stats.staminaRegenRate * Time.deltaTime;

            stats.currentStamina = Mathf.Clamp(stats.currentStamina, 0f, stats.maxStamina);
        }

        public void TakeDamage(float damage)
        {
            if (stats.currentArmor > 0)
            {
                float armorDamage = Mathf.Min(damage * 0.5f, stats.currentArmor);
                stats.currentArmor -= armorDamage;
                damage -= armorDamage;
            }

            stats.currentHealth -= damage;
            stats.currentHealth = Mathf.Max(stats.currentHealth, 0f);

            // Show damage indicator
            DamageIndicator damageIndicator = FindObjectOfType<DamageIndicator>();
            if (damageIndicator != null)
                damageIndicator.ShowDamage();

            if (stats.currentHealth <= 0)
                Die();
        }

        private void Die()
        {
            Debug.Log("Player died!");
            // TODO: Implement death logic
        }

        public void Heal(float amount)
        {
            stats.currentHealth += amount;
            stats.currentHealth = Mathf.Min(stats.currentHealth, stats.maxHealth);
        }

        public void RestoreArmor(float amount)
        {
            stats.currentArmor += amount;
            stats.currentArmor = Mathf.Min(stats.currentArmor, stats.maxArmor);
        }

        // Getters for pickup system
        public float GetCurrentHealth() => stats.currentHealth;
        public float GetMaxHealth() => stats.maxHealth;
        public float GetCurrentArmor() => stats.currentArmor;
        public float GetMaxArmor() => stats.maxArmor;
    }
}
